aten/src/ATen/native/cuda/Reduce.cuh

```diff
@@ -331,22 +356,38 @@ struct ReduceOp {
     return value;
   }
 
-  C10_DEVICE arg_t warp_reduce(arg_t value) const {
-    for (int offset = 1; offset < warpSize; offset <<= 1) {
+  C10_DEVICE arg_t block_x_reduce(arg_t value, char* shared_memory) const {
+    int dim_x = blockDim.x;
+    arg_t* shared = (arg_t*)shared_memory;
+    if (dim_x > warpSize) {
+      int address_base = threadIdx.x + threadIdx.y*blockDim.x;
+      shared[address_base] = value;
+      for (int offset = dim_x/2; offset >= warpSize; offset >>= 1) {
```

Comment:
`offset /= 2` seems clearer to me, but it's up to you.

Reply:
`>>=` is more efficient than `/=` (not that it's a deal breaker tho)
I'm always nervous with division in device code.